File: src/main/java/personal/ffivquizbot/slash_command/SlashCommandListener.java
package personal.ffivquizbot.slash_command;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.springframework.stereotype.Service;
import personal.ffivquizbot.slash_command.introduction.IntroductionCommandHandler;
import personal.ffivquizbot.slash_command.job_question.JobQuestionCommandHandler;

import java.util.Arrays;
import java.util.NoSuchElementException;

@Service
@RequiredArgsConstructor
@Slf4j
public class SlashCommandListener extends ListenerAdapter {

    private final JobQuestionCommandHandler jobQuestionCommandHandler;
    private final IntroductionCommandHandler introductionCommandHandler;

    @Override
    public void onSlashCommandInteraction(SlashCommandInteractionEvent event) {

        String name = event.getName();

        SlashCommandHandler targetCommandHandler = Arrays.asList(jobQuestionCommandHandler, introductionCommandHandler).stream()
                .filter(slashCommandHandler -> slashCommandHandler.getCommand().getCommandName().equals(name))
                .findFirst()
                .orElseThrow(NoSuchElementException::new);

        log.info("/ 커맨드를 실행합니다., 커맨드 이름 = {}", name);
        targetCommandHandler.handleCommand(event);
    }
}

File: src/main/kotlin/personal/ffivquizbot/slashcommand/SlashCommandListener.kt
package personal.ffivquizbot.slashcommand

import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import okhttp3.internal.immutableListOf
import org.springframework.stereotype.Service
import personal.ffivquizbot.slashcommand.astrologianskills.AstrologianSkillQuestionCommandHandler
import personal.ffivquizbot.slashcommand.introduction.IntroductionCommandHandler
import personal.ffivquizbot.slashcommand.invincibility.InvincibilityQuestionCommandHandler
import personal.ffivquizbot.slashcommand.jobquestion.JobQuestionCommandHandler

@Service
class SlashCommandListener(
    private val jobQuestionCommandHandler: JobQuestionCommandHandler,
    private val introductionCommandHandler:IntroductionCommandHandler,
    private val astrologianSkillQuestionCommandHandler: AstrologianSkillQuestionCommandHandler,
    private val invincibilityQuestionCommandHandler: InvincibilityQuestionCommandHandler,
) :ListenerAdapter() {

    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
        val name = event.name

        val handler = immutableListOf(
            jobQuestionCommandHandler,
            introductionCommandHandler,
            astrologianSkillQuestionCommandHandler,
            invincibilityQuestionCommandHandler,
        ).firstOrNull { handler -> handler.command.commandName == name }

        if (handler == null) {
            throw IllegalArgumentException("지원하지 않는 명령어가 입력되었습니다, ${name}")
        }

        handler.handleCommand(event)
    }
}